### PR TITLE
Moving feature-extraction pipeline to new testing scheme

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@LysandreJik

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/feature_extraction.py [hunk 1]
```diff
@@ -79,4 +79,10 @@ def __call__(self, *args, **kwargs):
         Return:
             A nested list of :obj:`float`: The features computed by the model.
         """
-        return super().__call__(*args, **kwargs).tolist()
+        results = super().__call__(*args, **kwargs)
+        if hasattr(results, "tolist"):
+            results = results.tolist()
+        else:
+            print("Sequential results")
```
Comment (sgugger) [inline src/transformers/pipelines/feature_extraction.py:86]: This should use the internal logger insread.

File: src/transformers/pipelines/feature_extraction.py [hunk 2]
```diff
@@ -79,4 +79,10 @@ def __call__(self, *args, **kwargs):
         Return:
             A nested list of :obj:`float`: The features computed by the model.
         """
-        return super().__call__(*args, **kwargs).tolist()
+        results = super().__call__(*args, **kwargs)
+        if hasattr(results, "tolist"):
+            results = results.tolist()
+        else:
```
Comment (sgugger) [inline src/transformers/pipelines/feature_extraction.py:85]: Test for a list/tuple and then raise a `ValueError` if we are not in that case? Seems a bit cleaner.

File: src/transformers/pipelines/base.py [hunk 3]
```diff
@@ -746,29 +746,33 @@ def _parse_and_tokenize(
...
 
-    def __call__(self, *args, **kwargs):
-        inputs = self._parse_and_tokenize(*args, **kwargs)
-        return self._forward(inputs)
+    def __call__(self, inputs, *args, **kwargs):
+        try:
+            model_inputs = self._parse_and_tokenize(inputs, *args, **kwargs)
+            outputs = self._forward(model_inputs)
+            return outputs
+        except ValueError:
```
Comment (sgugger) [inline src/transformers/pipelines/base.py:763]: Not of the scope of this PR, but more generally: I think we should have a proper `PaddingTokenMissingError` (or a better name) that we raise in this case, instead of relying on a generic `ValueError` that could come from something else.

File: tests/test_pipelines_common.py [hunk 4]
```diff
@@ -29,7 +29,11 @@
 
 
 def get_checkpoint_from_architecture(architecture):
-    module = importlib.import_module(architecture.__module__)
+    try:
+        module = importlib.import_module(architecture.__module__)
+    except Exception:
```
Comment (sgugger) [inline tests/test_pipelines_common.py:34]:
```suggestion
    except ImportError:
```
Let's try to have something not too broad here.

File: tests/test_pipelines_common.py [hunk 5]
```diff
@@ -46,8 +50,12 @@ def get_tiny_config_from_class(configuration_class):
     model_type = configuration_class.model_type
     camel_case_model_name = configuration_class.__name__.split("Config")[0]
 
-    module = importlib.import_module(f".test_modeling_{model_type.replace('-', '_')}", package="tests")
-    model_tester_class = getattr(module, f"{camel_case_model_name}ModelTester", None)
+    try:
+        module = importlib.import_module(f".test_modeling_{model_type.replace('-', '_')}", package="tests")
+        model_tester_class = getattr(module, f"{camel_case_model_name}ModelTester", None)
+    except Exception:
```
Comment (sgugger) [inline tests/test_pipelines_common.py:56]:
```suggestion
    except ImportError, AttributeError:
```
Same here.

File: tests/test_pipelines_feature_extraction.py [hunk 6]
```diff
@@ -14,13 +14,90 @@
...
+        # N = number of sequences passed
+        # T = number of tokens in sequence (depends on the tokenizer)
+        # E = number of embedding dimensions
+
+        shape = self.get_shape(outputs)
+        self.assertEqual(shape[0], 1)
+        # self.assertIsInstance(outputs[0], list)
+        # self.assertIsInstance(outputs[0][0], list)
+        # self.assertIsInstance(outputs[0][0][0], float)
+        # self.assertTrue(all(isinstance(el, float) for row in outputs for col in row for el in col))
```
Comment (sgugger) [inline tests/test_pipelines_feature_extraction.py:93]: What are all those comments? Something to clean up or fix?

File: tests/test_pipelines_feature_extraction.py [hunk 7]
```diff
@@ -14,13 +14,90 @@
...
+
+        outputs = feature_extractor(["This is a test", "Another test"])
+        shape = self.get_shape(outputs)
+        self.assertEqual(shape[0], 2)
+        # self.assertIsInstance(outputs, list)
+        # self.assertEqual(len(outputs), 2)
+        # self.assertIsInstance(outputs[0], list)
+        # self.assertIsInstance(outputs[0][0], list)
+        # self.assertIsInstance(outputs[0][0][0], float)
+        # self.assertTrue(all(isinstance(el, float) for row in outputs for col in row for el in col))
```
Comment (sgugger) [inline tests/test_pipelines_feature_extraction.py:103]: Same here.

File: tests/test_pipelines_feature_extraction.py [hunk 8]
```diff
@@ -14,13 +14,90 @@
...
+            subshapes = [self.get_shape(in_, shape) for in_ in input_]
+            if all(s == 0 for s in subshapes):
+                shape.append(len(input_))
+            else:
+                subshape = subshapes[0]
+                shape = [len(input_), *subshape]
+        elif isinstance(input_, float):
+            return 0
+        else:
+            raise Exception("We expect lists of floats, nothing else")
```
Comment (sgugger) [inline tests/test_pipelines_feature_extraction.py:65]:
```suggestion
            raise ValueError("We expect lists of floats, nothing else")
```

**Review by sgugger — APPROVED** [review]
Thanks for the PR! Left a few comments (mostly around the exceptions). It might be worth waiting for @LysandreJik to be back since he worked a lot on this with you, but if it's urgent we can merge and address his comments later on.

**Review by stas00 — COMMENTED** [review]
I'm trying to wrap my head around the from_pretraining chunk before your proposed edits:

So:

```
        if not hasattr(model, cls.base_model_prefix) and has_prefix_module:
            start_prefix = cls.base_model_prefix + "."
        if hasattr(model, cls.base_model_prefix) and not has_prefix_module:
            model_to_load = getattr(model, cls.base_model_prefix)
```

is really:

```
        if     (not hasattr(model, cls.base_model_prefix) and has_prefix_module):
            start_prefix = cls.base_model_prefix + "."
        if not (not hasattr(model, cls.base_model_prefix) and has_prefix_module):
            model_to_load = getattr(model, cls.base_model_prefix)
```

which is a clear if/else:

```
        if not hasattr(model, cls.base_model_prefix) and has_prefix_module:
            start_prefix = cls.base_model_prefix + "."
        else:
            model_to_load = getattr(model, cls.base_model_prefix)
```

is it not?

Those 2 conditions are complete opposites via (NOT), and thus shouldn't be separate. I think it'd make it easier to read/fix things.

**no, my bad - this is completely bogus boolean logic. The correct logic is `not (a and b)` is `not a or not b`. So please ignore my comment.**

File: src/transformers/modeling_utils.py [hunk 9]
```diff
@@ -1485,13 +1484,12 @@ def load(module: nn.Module, prefix=""):
 
         # Make sure we are able to load base models as well as derived models (with heads)
         start_prefix = ""
-        model_to_load = model
         if not hasattr(model, cls.base_model_prefix) and has_prefix_module:
             start_prefix = cls.base_model_prefix + "."
+        load(model, prefix=start_prefix)
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1489]: Why is this added here? In the case the following test is positive, we end up loading the model twice.

Comment (stas00) [reply src/transformers/modeling_utils.py:1489]: It seems that `from_pretrained` ignores the HEAD weights in some models, e.g.:

```
python -c 'from transformers import AutoModelForMaskedLM; model = AutoModelForMaskedLM.from_pretrained("Narsil/tiny-distilbert"); print(model)'

Some weights of the model checkpoint at Narsil/tiny-distilbert were not used when initializing DistilBertForMaskedLM: ['qa_outputs.weight', 'classifier.weight', 'pre_classifier.bias', 'qa_outputs.bias', 'classifier.bias', 'pre_classifier.weight']
[...]
```

Comment (stas00) [reply src/transformers/modeling_utils.py:1489]: ah, @Narsil, this is because you used the wrong module, this works correctly:
```
python -c 'from transformers import DistilBertForSequenceClassification; model = DistilBertForSequenceClassification.from_pretrained("Narsil/tiny-distilbert"); print(model)'
```

Comment (Narsil) [reply src/transformers/modeling_utils.py:1489]: This is exactly what failed actually. Bit of context, I used this script: https://gist.github.com/Narsil/7198efc17ad5c9bc4f711ffb0e0e3758 based on the original from @LysandreJik .

I realized there was an issue where the weights where incorrect (indeed we can't have 1 model for every possible head since at least ForMultipleChoice and ForSequenceClassification share the module `classifier` yet have different shapes (dim, 1) vs (dim, num_labels)).

I don't really know, but I ended up with a wrong model without seeing any error. Can't reproduce today for some reason.

Comment (Narsil) [reply src/transformers/modeling_utils.py:1489]: (When size mismatched, instead of loading the head or crashing, I ended with a random head in the pipeline leading to random results)

File: src/transformers/pipelines/base.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/pipelines/base.py:765]: Agreed.

File: src/transformers/pipelines/feature_extraction.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/pipelines/feature_extraction.py:86]: I'll remove it.

File: src/transformers/pipelines/feature_extraction.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/feature_extraction.py:85]: No it's linked to sequential vs batched, so both are valid.

I will invert both cases to avoid `hasattr` but I want to avoid code specific to the type of tensor that's the reason it's written this way.

File: tests/test_pipelines_feature_extraction.py (hunk 6 above)
Comment (Narsil) [reply tests/test_pipelines_feature_extraction.py:87]: Clean up. Some models return different shapes so we cannot rely on the output shape.
I'll remove.

File: src/transformers/modeling_utils.py (hunk 9 above)
Comment (Narsil) [reply src/transformers/modeling_utils.py]: Okay, to reproduce:

```python
DistilBertForSequenceClassification.from_pretrained("Narsil/tiny-distilbert-sequence-classification").classifier.weight
```
`python -c 'from transformers import *; print(DistilBertForSequenceClassification.from_pretrained("Narsil/tiny-distilbert-sequence-classification").classifier.weight)'`  (for the lazy like me)

Run that line a couple of times, and you will have different weights, even though the state_dict actually contains the head.

Comment (Narsil) [reply src/transformers/modeling_utils.py]: (And there is no log whatsoever)

**Comment by Narsil** [comment]
@sgugger It's not that it's urgent, it's that there's a quite big backlog (all pipelines tests need to be converted, and the goal is to start working on proper pipeline iteration (enabling more speed on GPU)

File: src/transformers/modeling_utils.py (hunk 9 above)
Comment (sgugger) [reply src/transformers/modeling_utils.py]: It looks like that particular model has not been saved properly: its state_dict contains keys that do not begin `distilbert`, as if it was an model without head, but still contains keys for a head. That's what makes the `from_pretrained` method confused but the problem is not in that method, it's this model state dict that is corrupted.

I double-checked but saving a `DistilBertForSequenceClassification` with either `save_pretrained` or `torch.save(model.state_dict())` will produce a usable state dict so I'm not sure how you obtained this one.

Comment (Narsil) [reply src/transformers/modeling_utils.py]: Ok, makes sense, I used the script made by @LysandreJik but I guess there was an issue, I'll fix it there then.

Comment (Narsil) [reply src/transformers/modeling_utils.py]: @sgugger the fact that it's not saying that it's missing keys is a bug though, no ? (The script had checks in place to make sure there were no missing keys, but nothing is reported)

Comment (Narsil) [reply src/transformers/modeling_utils.py]:
```python
from transformers import *

model = DistilBertForSequenceClassification.from_pretrained(None, config=DistilBertConfig(), state_dict = {})
model.state_dict().keys()    # "classifier.weight" and not "distilbert.classifier.weight"
model.save_pretrained("tmp")  # Same in the pytorch dict.
```
Let's move this discussion online so it's an easier back and forth :)

Comment (sgugger) [reply src/transformers/modeling_utils.py]: Yes, the wrong state dict definitely confused the `from_pretrained` function but there is a bug to fix indeed!

**Comment by Narsil** [comment]
I will merge this if that's OK (So I can prepare the next PR)

**Comment by sgugger** [comment]
It's fine by me, as long as comments from @LysandreJik (if any) are addressed later on and if you could double-check on the branch of #12939 that the new models used in those tests can be instantiated without any problem.
